### ISSUE TITLE
Determining `requires_grad` automatically

### ISSUE THREAD

**Issue by netheril96** [pytorch/pytorch#5349]
- OS: Ubuntu 16.04
- PyTorch version: 0.3.1
- How you installed PyTorch (conda, pip, source): pip
- Python version: Python 3.5.2
- CUDA/cuDNN version: 9.0/7.0
- GPU models and configuration:
- GCC version (if compiling from source):

Suppose we are training a GAN, with a very simple generator G(z) = tanh(W₁z + b₁) and discriminator D(x) = W₂x + b₂. The loss is a function of D(G(z)), and when calling `loss.backward()`, all of the gradients in the network are computed. However, during GAN training, the generator and discriminator are trained separately in alternation, therefore

1. When training D, the gradients do not need to back propagate to G.
2. When training G, we need only ∂D/∂x, but not ∂D/∂W₂ or ∂D/∂b₂.

The larger the network, the more unnecessary computation is done if we do not set `requires_grad` and `volatile` appropriately. According to my observation, few people understand the need to do this in PyTorch, and if they do, they will likely realize point (1) but miss point (2). If the network is even more complex (e.g. ACGAN + WGAN-GP), it will be even harder for a human to analyze.

TensorFlow does not have the same problem, because its optimizers receive a list of variables to optimize, and with that information they can determine exactly what computation are needed.

**Comment by apaszke** [comment]
TensorFlow assumes that it has much more knowledge (basically full knowledge) about the user code than we do. There's no way for us to determine automatically which Variables should be compute or not (unless you take care of annotating them for us), because we don't know which optimizer will you call next/in some time. Thanks for the proposal but I don't think this is feasible. If you have any concrete implementation ideas please reopen/let us know.

**Comment by netheril96** [comment]
It is actually feasible for PyTorch. By the time the user wants to compute the gradients, the dynamic graph has already been constructed, at which time the dependency information can be inferred from the graph. PyTorch cannot automatically infer what tensors not to compute during forward pass, but it certainly can during backward pass.

**Comment by apaszke** [comment]
Not really. Remember that optimizers aren't the only consumers of gradients. One can even roll out their own optimization methods manually.

**Comment by ezyang** [comment]
@apaszke Given that `no_grad` is coming, wouldn't it be feasible to also extend this with `only_grad(list of variables)` which has the same effect as setting `requires_grad=True` for the listed variables, and False for everything else? Then an optimizer can, in fact, "just take care of it" for you.

**Comment by apaszke** [comment]
How is that different than `torch.autograd.grad(out, just_those_vars)`?

**Comment by ezyang** [comment]
You would get to avoid building graph for `just_those_vars`. It would probably also make it simpler to design a new context-manager API to optimizer which takes care fo this for you.

**Comment by apaszke** [comment]
I'm just not sure what the exact semantics are, and how would you implement that. An example corner case: what if part of the graph involving other variables has already been built, and is used in the block?

**Comment by ezyang** [comment]
So... maybe the thing we're missing is a `torch.autograd.backward` variant that takes both inputs and outputs (like `torch.autograd.grad`) but accumulates gradients imperatively in `grad`. Then this would be a drop-in replacement at any point you want to call `loss.backward()` but train only certain parameters. The context manager is just icing on the API cake (for conveniently transforming `loss.backward()` calls into the elaborated version with explicit inputs).

Avoiding creating graphs in all cases seems difficult, because there's no convenient way to tell what leaves contributed to the history of a variable. I suppose you can make a modest improvement by *only* clipping the graph when a parameter is directly used, and conservatively build graphs for any variables that come in requiring gradient. This would still be a win if most of your graph management happens inside the context manager.

**Comment by apaszke** [comment]
Well this makes the semantics of `requires_grad` yet more complicated than they are now, because there's an extra way you can affect them at a distance, and it's much more complicated than a `no_grad` mode - it's not only checking a global flag, but if a Variable is a leaf or not, and if it is a leaf, then if it is in the "want grad" set. I don't think it's worth the complexity.

**Comment by ezyang** [comment]
Well, how about `torch.autograd.backward` with support for specifying which inputs you actually want to compute gradients for? I think it's pretty easy to polyfill this functionality (take each of the returned grads from `torch.autograd.grad` and manually add them to the `x.grad` of each input `x`) but it seems like a glaring omission from the current API.

**Comment by apaszke** [comment]
There's even a simpler way to implement this, using the logic that extracts the subgraph necessary to compute `torch.autograd.grad`. I think I'm ok with adding it, no strong preferences really. But I don't think we should have any context managers for this.

**Comment by cossio** [comment]
Any news on this after almost two years?

**Comment by albanD** [comment]
Hi,

This is not really high priority as there are easy ways to do the same thing with the current api.
As mentioned above, you can implement this in 3 lines with `autograd.grad()` and filling the `.grad` fields on the inputs manually.

But we would be happy to accept PR if you want to add an "inputs" argument to `autograd.backward()`.

**Comment by cossio** [comment]
The initial post in this thread illustrates the issue with GAN training. I am beginning to read the PyTorch tutorial (https://pytorch.org/tutorials/beginner/dcgan_faces_tutorial.html) and was wondering if the unnecessary gradient computation issue is relevant in that documentation page?

**Comment by albanD** [comment]
This is now in master, see the `inputs=` kwarg for `.backward()`.
